feat: Add an Autodiff Plugin

benjaminhuth opened the pull request:
This PR proposes a Plugin, which allows to do the transport of the covariance matrix with methods of automatic differentiation instead of computing the derivatives by hand (I presented this in an ACTS meeting a few weeks ago). Essentially, it provides the `AutodiffWrapperExtension` which can be used with templated extension in the `EigenStepper`. The necessary templated extensions (slightly changed copies of the extensions in Core/Propagator) can also be found in the plugin-folder.

The PR introduces:
- A new header-only plugin
- A new third-party dependency (autodiff)
- Integration tests for the AutodiffPropagator


PS: To explain the close and re-open of the PR, I misclicked in the first try when opening this...

paulgessinger commented:
Hey @benjaminhuth, thanks for this PR! 

I suppose @FabianKlimpel should have a look at this.

paulgessinger commented on thirdparty/README.md:
```diff
@@ -28,3 +28,18 @@ following folders removed:
+
+A copy of [autodiff](https://github.com/benjaminhuth/autodiff) revision f4a82c2.This 
+fork differs from the [original library](https://github.com/autodiff/autodiff) by 
+supporting `std::hypot` and some small changes to prevent the library from 
```
Hm, would there be a way to upstream these changes? I'm not familiar with the library, but I have a bit of a bad feeling of vendoring a fork that we might not be able to update.

benjaminhuth replied:
Yes, that's probably not the best way. I did this, because I up to now only added support for `std::hypot` in the corner-case needed in ACTS. I think to bring this into autodiff, I would need to generalize this (which should not be a big deal). So the question is if we wait with the PR until then or change this later?

paulgessinger commented on Plugins/Autodiff/include/Acts/Plugins/Autodiff/GenericDefaultExtension.hpp:
```diff
@@ -0,0 +1,241 @@
+/// D of the RKN4 stepping. This is a pure implementation by textbook.
+/// @note This is the templated generic type. The default one with 'double'
+/// is a typedef below.
+template <typename float_t>
+struct GenericDefaultExtension {
+  /// @brief Default constructor
+  GenericDefaultExtension() = default;
```
Do I understand correctly that this is a copy of the core default extension, where you 'only' add a floating point type, so you can plug in the special types from autodiff.

Would there be a way to not have to copy the class(es) here? Do you make other changes to the class other than the template parameter?

benjaminhuth replied:
Yeah, I also thought about just changing the files in the Core modules, but there are some changes that are I think not common in rest of ACTS:
- I use custom typedefs for Eigen types, like:
```c++
  template <typename T>
  using Vector3D = Eigen::Matrix<T, 3, 1>;
```
- To support some function which rely on `float` (like `computeEnergyLossMean(...)`) I had to do some a bit unusual things like `static_cast<float>(static_cast<double>(qop))`.

So these changes would bring a few inconsitencies with the rest of ACTS into this... But in principle it should be no problem to change the original files.

benjaminhuth commented:
I brought my changes (`std::hypot`, ...) now to the third party dependency autodiff, so there is no longer a dependency to my fork, but to the official one. 
Regarding the issue with the copied files for the Extensions, I'm not sure what's the way to go here. Is templating the classes in the Core-Module an option?

msmk0 commented:
Hi @benjaminhuth,

thanks for getting your changes upstream. From a first look, I can image that we could add relevant modifications to the Core. I will have a closer look tomorrow; also regarding the correct integration of `autodiff`.

benjaminhuth commented:
I removed the copies of the extensions in the Plugin-folder, and templated the extensions in the Core Module. So far everything works as expected. Still, I have some thoughts on this:

* Should the `GenericDefaultExtension` maybe be in `namespace detail` or something like that? I thought about API etc...
* Until now, I have put the integration tests in a separate directory. But then it might happen, that the integration tests evolve somehow, and these separated tests are forgotten. So, one could include the tests also in the main test-files, separated with `#ifdef AUTODIFF_PLUGIN` ... `#endif`. Maybe it is also not necessary to do all possible comparison tests with autodiff...

But appart from this, from my side there is nothing to add.

msmk0 reviewed (approved):
Thanks @benjaminhuth for updating this PR. I am happy with the conceptual changes; I suggested a few changes so the code style and naming is consistent with the rest of the code base.

There are a few other things that need to be fixed before this can go in:
- There should be a corresponding CMake option `ACTS_BUILD_PLUGIN_AUTODIFF` defined in the main `CMakeLists` that is also described in the documentation.
- The autodiff library is included in public header files and needs to be installed as well. We also need to be able to handle the case, where the library is already available on the system and the bundled version must not be used (see e.g. how this is handled for the `nlohmann::json` library). This does not seem to be handled correctly by the upstream CMake configuration so we either need to have a custom `CMakeLists.txt` for this or try to add a patch upstream.

msmk0 reviewed (changes requested):
Thanks @benjaminhuth for updating this PR. I am happy with the conceptual changes; I suggested a few changes so the code style and naming is consistent with the rest of the code base.

There are a few other things that need to be fixed before this can go in:
- There should be a corresponding CMake option `ACTS_BUILD_PLUGIN_AUTODIFF` defined in the main `CMakeLists` that is also described in the documentation.
- The autodiff library is included in public header files and needs to be installed as well. We also need to be able to handle the case, where the library is already available on the system and the bundled version must not be used (see e.g. how this is handled for the `nlohmann::json` library). This does not seem to be handled correctly by the upstream CMake configuration so we either need to have a custom `CMakeLists.txt` for this or try to add a patch upstream.

benjaminhuth commented:
Okay, I have changed the things @msmk0 mentioned. Actually most of the things I read in the contribution guide, but obviously forgot them.

Regarding the CMake integration of the autodiff library, I changed the following things:
* Now there exists a `ACTS_USE_SYSTEM_AUTODIFF` option
* When disabled, the `autodiff` library should install itself the same way as `nlohman::json` in the include folder of the install path
* The only thing I changed in `autodiff`'s `CMakeLists.txt` is that I commented out the `add_subdirectory(tests)` and `add_subdirectory(examples)` (these folders are removed), and it seems to work well. Is this an acceptable difference from the original library?

I hope this addresses the mentioned problems.

msmk0 commented on Tests/IntegrationTests/CMakeLists.txt:
```diff
@@ -42,4 +42,4 @@ add_integrationtest(PropagationCompareEigenStraightLine PropagationCompareEigenS
 add_subdirectory_if(Fatras ACTS_BUILD_FATRAS)
 add_subdirectory_if(Legacy ACTS_BUILD_PLUGIN_LEGACY)
-add_subdirectory_if(Autodiff ACTS_BUILD_AUTODIFF_PLUGIN)
+add_subdirectory_if(Autodiff ACTS_BUILD_PLUGIN_AUTODIFF)
```
Please keep this in alphabetical order.

msmk0 commented on thirdparty/README.md:
```diff
@@ -31,13 +31,12 @@ following folders removed:
 
-A copy of [autodiff](https://github.com/autodiff/autodiff), revision eefd2f7.
-The following files/folders have been removed:
+A copy of [autodiff](https://github.com/autodiff/autodiff), revision 430a0f3.
```
Is this a specific commit or the tagged version `v0.5.11`. If it is the later, I would prefer to mention the tagged version here.

msmk0 commented on thirdparty/README.md:
```diff
@@ -31,13 +31,12 @@ following folders removed:
--   test
--   most files except headers
+-   CMakeLists.txt
+-   .clang-format
```
Is the clang-format config necessary? If not, please remove it.

msmk0 commented on Plugins/CMakeLists.txt:
```diff
@@ -7,7 +7,7 @@ add_component_if(Identification PluginIdentification ACTS_BUILD_PLUGIN_IDENTIFIC
 add_component_if(Legacy PluginLegacy ACTS_BUILD_PLUGIN_LEGACY)
 add_component_if(Sycl PluginSycl ACTS_BUILD_PLUGIN_SYCL)
-add_component_if(Autodiff AutodiffPlugin ACTS_BUILD_AUTODIFF_PLUGIN)
+add_component_if(Autodiff PluginAutodiff ACTS_BUILD_PLUGIN_AUTODIFF)
```
Please keep this in alphabetical order.

msmk0 commented on Plugins/Autodiff/CMakeLists.txt:
```diff
@@ -1,24 +1,25 @@
+    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include/>
+    $<INSTALL_INTERFACE:include>)
+ 
+# the target name is different when using find_package(...) or add_subdirectory(...)
+if(ACTS_USE_SYSTEM_AUTODIFF)
+    set(AUTODIFF_TARGET autodiff::autodiff)
+else()
+    set(AUTODIFF_TARGET autodiff)
+endif()
```
Suggestion: add the missing `autodiff::autodiff` target as an alias target via `add_library(<name> ALIAS <target>)` when you include it via `add_subdirectory` in the main `CMakeLists.txt`. You can then always use the `autodiff::autodiff` target regardless of how the package is included. The plugin should not care whether we use the bundled version or not. It also means we have to do this only in one location if we e.g. decide to use the autodiff library somwhere else.

In the long term, we should try to get CMake fixes upstream so we do not need to patch this locally. But not for this PR.

msmk0 commented on Core/include/Acts/Propagator/DenseEnvironmentExtension.hpp:
```diff
@@ -28,38 +30,35 @@ namespace Acts {
 
   /// Momentum at a certain point
-  float_t currentMomentum = 0.;
+  scalar_t currentMomentum = 0.;
```
```suggestion
  Scalar currentMomentum = 0.;
```
Nitpicking: this uses the local typedef in other parts of the codebase.

msmk0 reviewed (changes requested):
Thanks. I have a few minor follow ups.

msmk0 commented on Plugins/Autodiff/include/Acts/Plugins/Autodiff/AutodiffExtensionWrapper.hpp:
```diff
@@ -0,0 +1,172 @@
+
+    // Init dependent values for autodiff
+    AutodiffFreeVector initial_params;
+    initial_params << stepper.position(state.stepping),
+        stepper.time(state.stepping), stepper.direction(state.stepping),
+        stepper.charge(state.stepping) / stepper.momentum(state.stepping);
```
@benjaminhuth Still needs to be fixed.

benjaminhuth commented on Core/include/Acts/Propagator/DenseEnvironmentExtension.hpp:
```diff
@@ -350,18 +363,21 @@ struct DenseEnvironmentExtension {
       g = -computeEnergyLossMean(slab, state.options.absPdgCode,
-                                 state.options.mass, qop[0]);
+                                 state.options.mass,
+                                 static_cast<double>(qop[0]));
```
I think this is not possible, because of the conversation rules. `autodiff::dual<T>` does support a (implicit or explicit) conversion to `T`, which is `double` in our case. A `static_cast<float>(...)` would fail. So we must apply a `static_cast<double>` as a generic method to get an ordinary floating point number from `double` as well as `autodiff::dual<double>` which are the two use cases we have. When not wanting an implicit conversion from `double` to `float`, we would need `static_cast<float>(static_cast<double>(...))`.

Another method would be a custom metafunction `scalar_to_float` which has spezialisations for `double` and `autodiff::dual<double>`.

just out of interest: Is there a specific reason why the energy loss computations are done with `float`?

msmk0 replied:
Thanks, I was suspecting something like this. I agree with you that `static_cast<double>(...)` is the simplest solution. Let's leave it like that.

All material interactions are done in `float` since there really is no need for higher precision. The material parameters are only stored in `float` both to keep them the size down and because they are usually not known with high precision anyways. E.g. the energy loss is just the mean of a statistical process and the statistical variations are much larger than the float precision. Computing them in double would be a waste of resources.

benjaminhuth commented on Core/include/Acts/Propagator/DenseEnvironmentExtension.hpp:
```diff
@@ -350,18 +363,21 @@ struct DenseEnvironmentExtension {
       g = -computeEnergyLossMode(slab, state.options.absPdgCode,
-                                 state.options.mass, qop[0]);
+                                 state.options.mass,
+                                 static_cast<double>(qop[0]));
```
See above

benjaminhuth commented on Core/include/Acts/Propagator/DenseEnvironmentExtension.hpp:
```diff
@@ -372,11 +388,13 @@ struct DenseEnvironmentExtension {
           dgdqopValue = deriveEnergyLossMeanQOverP(
-              slab, state.options.absPdgCode, state.options.mass, qop[0]);
+              slab, state.options.absPdgCode, state.options.mass,
+              static_cast<double>(qop[0]));
```
See above

benjaminhuth commented on Core/include/Acts/Propagator/DenseEnvironmentExtension.hpp:
```diff
@@ -372,11 +388,13 @@ struct DenseEnvironmentExtension {
           dgdqopValue = deriveEnergyLossModeQOverP(
-              slab, state.options.absPdgCode, state.options.mass, qop[0]);
+              slab, state.options.absPdgCode, state.options.mass,
+              static_cast<double>(qop[0]));
```
See above

msmk0 reviewed (approved):
@benjaminhuth Thanks for the final changes. There is nothing else to do from my side. I will request a review from @FabianKlimpel to check the changes to the propagator code and then this can go in.

In the meantime, could you merge the latest master branch?

FabianKlimpel commented on Plugins/Autodiff/include/Acts/Plugins/Autodiff/AutodiffExtensionWrapper.hpp:
```diff
@@ -0,0 +1,176 @@
+    out.segment<3>(eFreeDir0) = final_dir / final_dir.norm();
+
+    // qop
+    out(eFreeQOverP) = state.stepping.q / state.stepping.p;
```
Same here

FabianKlimpel commented on thirdparty/autodiff/autodiff/forward/eigen.hpp:
```diff
@@ -0,0 +1,344 @@
+struct NumTraits;
+
+template<typename T, typename G>
+struct NumTraits<autodiff::forward::Dual<T, G>> : NumTraits<autodiff::forward::ValueType<T>> // permits to get the epsilon, dummy_precision, lowest, highest functions
```
Could you split this line?

paulgessinger replied:
This is thirdparty code.

msmk0 replied:
I did not see that. Yes, of course, for thirdparty code it does not apply. This stays as-is.

FabianKlimpel commented on thirdparty/autodiff/autodiff/forward/forward.hpp:
```diff
@@ -0,0 +1,1807 @@
+struct ValueType<BinaryExpr<Op, L, R>> { using type = common<typename ValueType<plain<L>>::type, typename ValueType<plain<R>>::type>; };
+
+template<typename Op, typename L, typename C, typename R>
+struct ValueType<TernaryExpr<Op, L, C, R>> { using type = common<typename ValueType<plain<L>>::type, common<typename ValueType<plain<C>>::type, typename ValueType<plain<R>>::type>>; };
```
In this file are also some pretty long lines.

msmk0 replied:
I am surprised that this is not detected by clang-format?

FabianKlimpel replied:
I assume that either the autodiff or even the thirdparty folder isn't covered

paulgessinger replied:
Doesn't run on third part code.

paulgessinger replied:
This is one reason I believe we should not bundle third party code the way we do.

benjaminhuth replied:
In the first version of my Plugin I used CMake's `ExternalProject` module, to download the dependencies at configure time (but that lead to a problem on one of the test systems). The source files then are in the build-folder. Maybe this could be a cleaner way?

msmk0 replied:
If we want to change this (which I still think we should not), it needs to be done in a separate PR. Please leave this as-is for now.

msmk0 commented on Core/include/Acts/Propagator/DefaultExtension.hpp:
```diff
@@ -8,17 +8,22 @@
+#include <array>
+
 namespace Acts {
+namespace detail {
```
That would be a backward incompatible change.

FabianKlimpel replied:
So backward compatibility > consistency?

msmk0 replied:
If we break this, we would have to increase the major version. I do not think this is necessary to do for such an otherwise unrelated change.

benjaminhuth replied:
I'm not sure, maybe I don't understand this API-versioning thing correctly, but if `DefaultExtension.hpp` in fact only consists of
```c++
#include "blablabla/detail/GenericDefaultExtension.hpp"

namespace Acts {
    using DefaultExtension = detail::GenericDefaultExtension<double>;
}
```
why would that break the backwards-compatibility? The symbols in the `namespace Acts` stay as they where, right?

msmk0 replied:
Yes, that would work.

benjaminhuth replied:
Okay, I moved the implementation to the `details` directory, and left the typedefs where they where. Is that okay for everyone?

benjaminhuth commented on Plugins/Autodiff/include/Acts/Plugins/Autodiff/AutodiffExtensionWrapper.hpp:
```diff
@@ -0,0 +1,176 @@
+    initial_params.segment<3>(eFreePos0) = stepper.position(state.stepping);
+    initial_params(eFreeTime) = stepper.time(state.stepping);
+    initial_params.segment<3>(eFreeDir0) = stepper.direction(state.stepping);
+    initial_params(eFreeQOverP) =
+        stepper.charge(state.stepping) / stepper.momentum(state.stepping);
```
Hmm. I'm not exactly sure how this can be resolved... By an assertion or something like that?

FabianKlimpel replied:
Like that:
https://github.com/acts-project/acts/blob/41d6cbc643a0de4e0640f6637d80103f58a71c7e/Core/include/Acts/Propagator/EigenStepper.ipp#L63

benjaminhuth replied:
Okay, I think I resolved this.

benjaminhuth commented on Core/include/Acts/Propagator/DefaultExtension.hpp:
```diff
@@ -8,17 +8,22 @@
+/// @note This it templated on the floating point type because of the autodiff
+/// plugin.
+template <typename scalar_t>
+struct GenericDefaultExtension {
```
I did this when moving the files to the `detail` directory

msmk0 commented:
@paulgessinger Why do you think this is backward incompatible?

paulgessinger commented:
> @paulgessinger Why do you think this is backward incompatible?

I misread the discussion. Sorry

msmk0 commented on Plugins/Autodiff/include/Acts/Plugins/Autodiff/AutodiffExtensionWrapper.hpp:
```diff
@@ -0,0 +1,176 @@
+  using AutodiffFreeVector = ActsMatrix<AutodiffScalar, 8, 1>;
+
+  /// Instance of the base extension
+  basic_extension_t<double> m_doubleExtension;
```
@benjaminhuth Could you comment?

benjaminhuth replied:
You find extension templated with `AutodiffScalar` in line 143. The reason for this is hat you need the double-extension to communicate with the "outer world" (The stepper). You cannot do this computation with autodiff, since then we would need to template the stepper itself. So, from the outside the Wrapper behaves exactly like the underlying Extension with double, just internally it computes the jacobian with autodiff. Another advantage is, that with this structure we don't need to differentiate through the adaptive stepsize estimation in the stepper.
The `basic_extension_t<AutodiffScalar>` is just necessary locally, its not necessary that it is a class member. To work with it, you need the `FakeStepper` etc., defined in the private section of the class.
I hope this clarifies it a bit.

msmk0 replied:
Ok, thanks. That makes sense to me.

msmk0 reviewed (approved):
I'm happy w/ the changes. There is one open question from Fabian that I would ask @benjaminhuth to comment.

Edit: And please update the branch to the recent master.

msmk0 commented:
@benjaminhuth Could you comment on the last open question from @FabianKlimpel so we can merge this?

benjaminhuth commented:
It's about the question with the `basic_extension_t<double>`, right? I commented on it on Friday I think, but for some reason at least I can only see it in the files-changed-tab.

https://github.com/acts-project/acts/pull/384/files#r498673984

msmk0 commented:
Yes, that was the question. I did not see your answer from Friday. The Github interface is not very helpful there. I am happy with your explanation. I would suggest that you add that as a comment to that variable and then we merge it.

@FabianKlimpel Does that work for you?

FabianKlimpel commented:
> Yes, that was the question. I did not see your answer from Friday. The Github interface is not very helpful there. I am happy with your explanation. I would suggest that you add that as a comment to that variable and then we merge it.
> 
> @FabianKlimpel Does that work for you?

yes
